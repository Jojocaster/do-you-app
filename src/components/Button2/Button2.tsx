import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useRef } from 'react'
import { Animated, StyleSheet, TouchableWithoutFeedback } from 'react-native'
import Colors from '../../constants/Colors'
import useColorScheme from '../../hooks/useColorScheme'
import { Text, View } from '../Themed'

interface ButtonProps {
  children?: string
  elevated?: boolean
  iconSize?: number
  onPress?: () => void
  icon?: React.ComponentProps<typeof MaterialCommunityIcons>['name']
  variant?: 'xs' | 'sm' | 'md' | 'lg'
}

const Content: React.FC<
  Pick<ButtonProps, 'children' | 'icon' | 'variant' | 'iconSize'> & {
    visible?: boolean
  }
> = ({ children, icon, visible = true, variant = 'md', iconSize }) => {
  const theme = useColorScheme()
  const activeVariant = variants[variant]

  return (
    <View style={[styles.content, { opacity: visible ? 1 : 0 }]}>
      {icon && (
        <MaterialCommunityIcons
          name={icon}
          size={iconSize || activeVariant.text.fontSize}
          style={{
            marginRight: children ? 3 : 0,
            color: Colors[theme].button.text,
          }}
        />
      )}
      {children && (
        <Text
          style={[
            styles.text,
            activeVariant.text,
            { color: Colors[theme].button.text },
          ]}
        >
          {children}
        </Text>
      )}
    </View>
  )
}

export const Button2: React.FC<ButtonProps> = (props) => {
  const scale = useRef(new Animated.Value(1))
  const theme = useColorScheme()
  const { elevated, onPress, variant = 'md' } = props

  const activeVariant = variants[variant]

  const onPressIn = () => {
    Animated.spring(scale.current, {
      toValue: 0.8,
      friction: 4,
      useNativeDriver: true,
    }).start()
  }

  const onPressOut = () => {
    Animated.spring(scale.current, {
      toValue: 1,
      friction: 2,
      useNativeDriver: true,
    }).start()
  }

  return (
    <TouchableWithoutFeedback
      delayLongPress={0}
      onPress={onPress}
      onPressIn={onPress ? onPressIn : undefined}
      onPressOut={onPress ? onPressOut : undefined}
    >
      <Animated.View
        style={[styles.container, elevated ? styles.elevated : null]}
      >
        {/* TODO: clean up */}
        <Animated.View
          style={[
            styles.button,
            activeVariant.button,
            {
              transform: [{ scale: scale.current }],
              backgroundColor: Colors[theme].button.back,
            },
          ]}
        >
          <Content {...props} visible={false} />
          <View
            style={[
              styles.button,
              activeVariant.button,
              styles.inner,
              { backgroundColor: Colors[theme].button.background },
            ]}
          >
            <Content {...props} />
          </View>
        </Animated.View>
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const xsStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 8,
  },
  text: {
    fontSize: 8,
  },
})

const smStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 8,
  },
  text: {
    fontSize: 10,
  },
})

const mdStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 16,
  },
  text: {
    fontSize: 12,
  },
})

const lgStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 16,
    paddingVertical: 6,
  },
  text: {
    fontSize: 13,
  },
})

const variants = {
  xs: xsStyles,
  sm: smStyles,
  md: mdStyles,
  lg: lgStyles,
}

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    display: 'flex',
    paddingLeft: 4,
    paddingTop: 4,
    flexWrap: 'wrap',
  },
  button: {
    backgroundColor: 'white',
    borderRadius: 3,
    borderWidth: 1,
    borderBottomColor: 'black',
    paddingVertical: 4,
  },
  elevated: {
    elevation: 2,
    shadowOffset: {
      height: 3,
      width: 3,
    },
    shadowOpacity: 0.1,
  },
  // buttonSm: {
  //   paddingHorizontal: 8,
  // },
  // buttonLg: {
  //   paddingHorizontal: 16,
  // },
  inner: {
    position: 'absolute',
    top: -4,
    left: -5,
  },
  content: {
    backgroundColor: 'transparent',
    display: 'flex',
    alignItems: 'center',
    flexDirection: 'row',
  },
  text: {
    color: 'black',
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
  // textSm: {
  //   fontSize: 10,
  // },
  // textLg: {
  //   fontSize: 12,
  // },
})
