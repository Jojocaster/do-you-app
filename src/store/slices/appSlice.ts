import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ColorSchemeName } from 'react-native'
import { getEnv } from '../../constants/Env'

export interface RemoteTheme {
  name: string
  theme: ColorSchemeName | null
  icon: string | null
  playerImage: string | null
}
interface Config {
  assets: string
  defaultTheme: RemoteTheme
  themes: RemoteTheme[]
}
interface AppState {
  // used to manage "what's new" modals based on the version
  whatsNew: string | null
  config: Config | null
  // store last config update
  updatedAt: number | null
}

const initialState: AppState = {
  whatsNew: null,
  config: null,
  updatedAt: null,
}

export const fetchConfig = createAsyncThunk('app/fetchConfig', async () => {
  const url = getEnv('API_URL')
  const response = await fetch(`${url}/config`)
  const data = await response.json()

  return data as Config
})

const appSlice = createSlice({
  initialState,
  name: 'app',
  reducers: {
    updateWhatsNew(state, action: PayloadAction<string>) {
      state.whatsNew = action.payload
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action) => {
      console.log('[fetchConfig]: resolved')
      state.updatedAt = new Date().getTime()
      state.config = action.payload
    })
  },
})

export const { updateWhatsNew } = appSlice.actions
export default appSlice.reducer
