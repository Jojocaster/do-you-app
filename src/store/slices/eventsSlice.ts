import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'

export const fetchEvents = createAsyncThunk('events/fetchEvents', async () => {
  const response = await fetch(
    'https://mahina.app/mahina/app/55559291029?page=1'
  )
  const data = await response.json()

  return data as Events
})

interface EventDate {
  date: string
  day: string
  month: string
}
interface EventData {
  description: string
  id: number
  image: string
  location: string
  price_min: number
  price_max: number | null
  title: string
  ticket_links: {
    link: string
    domain: string
  }[]
  start_date: EventDate
  end_date: EventDate
}
interface Events {
  event_dates: string[]
  events_data: EventData[]
  loading: boolean
  updatedAt: number | null
}

const initialState: Events = {
  event_dates: [],
  events_data: [],
  loading: false,
  updatedAt: null,
}

const eventsSlice = createSlice({
  initialState,
  name: 'event',
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchEvents.fulfilled, (state, action) => {
      state.event_dates = action.payload.event_dates
      state.events_data = action.payload.events_data
      state.loading = false
      state.updatedAt = new Date().getTime()
    })
    builder.addCase(fetchEvents.pending, (state) => {
      state.loading = true
    })
    builder.addCase(fetchEvents.rejected, (state) => {
      state.loading = false
    })
  },
})

export default eventsSlice.reducer
