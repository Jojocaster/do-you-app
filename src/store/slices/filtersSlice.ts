import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { ARCHIVES_URL } from '../../constants/Endpoints'

export interface Filter {
  id: number
  name: string
}

interface Filters {
  shows: Filter[]
  loading: boolean
  error: boolean
  lastUpdated: number | null
}

export const fetchFilters = createAsyncThunk(
  'filters/fetchFilters',
  async () => {
    const response = await fetch(`${ARCHIVES_URL}/hosts/page/1/size/60/`)
    const data = await response.json()
    return data as Filter[]
  }
)

const initialState: Filters = {
  shows: [],
  loading: false,
  error: false,
  lastUpdated: null,
}

const filtersSlice = createSlice({
  initialState,
  name: 'filters',
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchFilters.pending, (state) => {
      state.loading = true
      state.error = false
    }),
      builder.addCase(fetchFilters.fulfilled, (state, action) => {
        state.loading = false
        state.shows = action.payload
        state.error = false
        state.lastUpdated = new Date().getTime()
      }),
      builder.addCase(fetchFilters.rejected, (state) => {
        state.loading = false
        state.error = true
      })
  },
})

export default filtersSlice.reducer
