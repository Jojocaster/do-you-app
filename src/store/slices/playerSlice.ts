import { createSlice, PayloadAction } from '@reduxjs/toolkit'

export enum PlayerStatus {
  PLAYING,
  LOADING,
  PAUSED,
  BUFFERING,
}

const initialState = {
  status: PlayerStatus.PAUSED,
}

const playerSlice = createSlice({
  initialState,
  name: 'player',
  reducers: {
    updatePlayerStatus(state, action: PayloadAction<PlayerStatus>) {
      state.status = action.payload
    },
  },
})

export const { updatePlayerStatus } = playerSlice.actions
export default playerSlice.reducer
