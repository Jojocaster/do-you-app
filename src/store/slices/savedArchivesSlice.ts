import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ArchiveItem } from '../../components/ArchivesList/ArchivesList.types'

interface SavedArchive {
  lastUpdated: number
  archive: ArchiveItem
}

const initialState: { archives: SavedArchive[] } = {
  archives: [],
}

const savedArchivesSlice = createSlice({
  initialState,
  name: 'savedArchives',
  reducers: {
    addArchive(state, action: PayloadAction<ArchiveItem>) {
      state.archives.push({
        lastUpdated: new Date().getTime(),
        archive: action.payload,
      })
    },
    deleteArchive(state, action: PayloadAction<ArchiveItem>) {
      const archives = state.archives.filter((a) => {
        return a.archive.id !== action.payload.id
      })

      state.archives = archives
    },
  },
})

export const { addArchive, deleteArchive } = savedArchivesSlice.actions
export default savedArchivesSlice.reducer
