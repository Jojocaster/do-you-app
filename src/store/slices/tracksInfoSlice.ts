import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TRACKLIST_URL } from '../../constants/Endpoints'

interface TrackInfo {
  album: string
  artist: string
  label: string
  played_datetime: string
  release_date: string
  song_link: string
  timecode: string
  title: string
}

export interface TracksInfoState {
  tracks: TrackInfo[]
  loading: boolean
  lastUpdated: number | null
}

export const fetchTracksInfo = createAsyncThunk(
  'tracksInfo/fetchShowInfo',
  async () => {
    const response = await fetch(TRACKLIST_URL)
    const data = await response.json()
    return data as { tracks: TrackInfo[] }
  }
)

const initialState: TracksInfoState = {
  lastUpdated: null,
  loading: true,
  tracks: [],
}

const tracksInfoSlice = createSlice({
  initialState,
  name: 'tracksInfo',
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      fetchTracksInfo.fulfilled,
      (state, action: PayloadAction<{ tracks: TrackInfo[] }>) => {
        state.tracks = action.payload.tracks
        state.loading = false
        state.lastUpdated = new Date().getTime()
      }
    ),
      builder.addCase(fetchTracksInfo.pending, (state) => {
        state.loading = true
      }),
      builder.addCase(fetchTracksInfo.rejected, (state) => {
        state.loading = false
        state.lastUpdated = new Date().getTime()
      })
  },
})

export default tracksInfoSlice.reducer
