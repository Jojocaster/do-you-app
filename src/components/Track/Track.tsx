import { MaterialIcons } from '@expo/vector-icons'
import { format, parseJSON } from 'date-fns'
import { useEffect, useMemo, useState } from 'react'
import { StyleSheet, TouchableOpacity } from 'react-native'
import Colors from '../../constants/Colors'
import useColorScheme from '../../hooks/useColorScheme'
import { TrackInfo } from '../../store/slices/tracksInfoSlice'
import { Button } from '../Button/Button'
import * as Clipboard from 'expo-clipboard'
import { Text, View } from '../Themed'

export const Track: React.FC<{
  active: boolean
  onToggle: (track: string) => void
  track: TrackInfo
}> = ({ active, track, onToggle }) => {
  const [copied, setCopied] = useState(false)
  const theme = useColorScheme()
  const timecode = parseJSON(track.played_datetime)
  const formattedTimecode = format(timecode, 'HH:mm')

  const onCopy = () => {
    try {
      Clipboard.setString(`${track.artist} - ${track.title}`)
    } catch (e) {
      console.log('There was an error copying to clipboard.')
    } finally {
      setCopied(true)
    }
  }

  useEffect(() => {
    setCopied(false)
  }, [active])

  // big performance impact, re-rendering all components added a ~200ms delay
  // useMemo prevents that by only re-rendering the tracks involved
  return useMemo(
    () => (
      <View>
        <View style={styles.container} key={track.played_datetime}>
          <View style={styles.timecode}>
            <Text
              style={{
                fontSize: 14,
              }}
            >
              {formattedTimecode}
            </Text>
          </View>
          <TouchableOpacity
            style={{ flex: 1 }}
            onPress={() => onToggle(track.played_datetime)}
          >
            <View style={{ display: 'flex', flexDirection: 'row' }}>
              <View style={{ marginRight: 5 }}>
                <MaterialIcons
                  name={active ? 'expand-less' : 'expand-more'}
                  size={16}
                  color={Colors[theme].text}
                />
              </View>
              <View style={styles.trackContainer}>
                <Text
                  style={{
                    color: Colors[theme].primary,
                    fontWeight: 'bold',
                    fontSize: 14,
                  }}
                >
                  {track.artist}
                </Text>
                <View>
                  <Text>{track.title}</Text>
                </View>

                <View>
                  {active && (
                    <View style={{ marginTop: 10, marginBottom: 10 }}>
                      <Text style={{ fontSize: 12 }}>
                        Confidence:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.score / 10}/10
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Album:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.album}
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Label:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.label}
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Release:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.release_date}
                        </Text>
                      </Text>
                      <View
                        style={styles.separator}
                        lightColor="#eee"
                        darkColor="rgba(255,255,255,0.1)"
                      />
                      <View style={{}}>
                        <Button
                          color={copied ? Colors[theme].primary : undefined}
                          icon={copied ? 'check' : 'content-copy'}
                          onPress={onCopy}
                        >
                          {copied ? `Copied` : 'Copy'}
                        </Button>
                      </View>
                    </View>

                    // <View
                    //   style={[
                    //     styles.codeHighlightContainer,
                    //     styles.homeScreenFilename,
                    //   ]}
                    //   darkColor="rgba(255,255,255,0.05)"
                    //   lightColor="rgba(0,0,0,0.05)"
                    // >
                    //   <Text>Album: {track.album}</Text>
                    //   <Text>Release: {track.release_date}</Text>
                    // </View>
                  )}
                </View>
              </View>
            </View>
          </TouchableOpacity>
        </View>
      </View>
    ),
    [active, theme, copied]
  )
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 10,
  },
  timecode: {
    flexBasis: 60,
  },
  trackContainer: {
    flex: 1,
  },
  codeHighlightContainer: {
    fontSize: 10,
    borderRadius: 3,
    padding: 20,
  },
  homeScreenFilename: {
    marginVertical: 10,
    display: 'flex',
  },
  separator: {
    marginVertical: 10,
    height: 1,
    width: '50%',
  },
})
