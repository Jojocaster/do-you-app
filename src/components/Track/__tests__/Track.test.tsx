import { MaterialCommunityIcons, MaterialIcons } from '@expo/vector-icons'
import { fireEvent } from '@testing-library/react-native'
import { Clipboard, TouchableOpacity } from 'react-native'
import renderer, { act } from 'react-test-renderer'
import { TrackInfo } from '../../../store/slices/tracksInfoSlice'
import { Button2 } from '../../Button2/Button2'
import { Track } from '../Track'

//@ts-ignore
const spy = jest
  .spyOn(Clipboard, 'setString')
  .mockImplementation(() => jest.fn())

const mockStore = {
  settings: {
    darkTheme: false,
  },
  tracksInfo: { loading: true, tracks: [] },
}

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn((cb) => cb(mockStore)),
}))

// mock for expo icons
jest.mock(
  '@expo/vector-icons/build/vendor/react-native-vector-icons/lib/create-icon-set.js',
  () => {
    return () => ''
  }
)

const mockTrack: TrackInfo = {
  album: 'Ema By The Sea',
  artist: 'Pink Shabab',
  label: 'MERLIN - Karaoke Kalk',
  played_datetime: '2022-07-29 10:46:32+0100',
  release_date: '2019-11-29',
  score: 100,
  song_link: 'https://lis.tn/WITYQZ',
  timecode: '00:22',
  title: 'Let Me Explain',
}

const mockToggle = jest.fn()

describe('Track', () => {
  it('should render', () => {
    const wrapper = renderer.create(
      <Track track={mockTrack} active={false} onToggle={mockToggle} />
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  describe('when active', () => {
    it('should show the track details', () => {
      const wrapper = renderer.create(
        <Track track={mockTrack} active={false} onToggle={mockToggle} />
      )
      const details = wrapper.root.findAllByProps({ testID: 'trackDetails' })
      expect(details.length).toEqual(0)

      act(() => {
        wrapper.update(
          <Track track={mockTrack} active={true} onToggle={mockToggle} />
        )
      })

      expect(wrapper.root.findByProps({ testID: 'trackDetails' })).toBeDefined()
    })

    it('should update the icon', () => {
      const wrapper = renderer.create(
        <Track track={mockTrack} active={false} onToggle={mockToggle} />
      )
      const icon = wrapper.root.findByType(MaterialIcons)
      expect(icon.props.name).toEqual('expand-more')

      act(() => {
        wrapper.update(
          <Track track={mockTrack} active={true} onToggle={mockToggle} />
        )
      })
      expect(icon.props.name).toEqual('expand-less')
    })
  })

  describe('when pressed', () => {
    it('should call `onPress`', () => {
      const wrapper = renderer.create(
        <Track track={mockTrack} active={false} onToggle={mockToggle} />
      )

      const btn = wrapper.root.findByType(TouchableOpacity)
      fireEvent.press(btn)
      expect(mockToggle).toHaveBeenCalledWith(mockTrack.played_datetime)
    })
  })

  describe('when copy is pressed', () => {
    it('should store the title in the clipboard', () => {
      const wrapper = renderer.create(
        <Track track={mockTrack} active={true} onToggle={mockToggle} />
      )
      const btn = wrapper.root.findByType(Button2)
      fireEvent.press(btn)

      expect(spy).toHaveBeenCalledWith(
        `${mockTrack.artist} - ${mockTrack.title}`
      )
    })
  })
})
