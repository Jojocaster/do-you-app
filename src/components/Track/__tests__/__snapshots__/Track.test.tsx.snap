// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track should render 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "display": "flex",
          "flexDirection": "row",
          "marginBottom": 10,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "flexBasis": 60,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            Object {
              "fontSize": 14,
            },
          ]
        }
      >
        10:46
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "marginRight": 5,
            },
          ]
        }
      >
        <Text />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                Object {
                  "color": "#3a5ed6",
                  "fontSize": 14,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Pink Shabab
          </Text>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  undefined,
                ]
              }
            >
              Let Me Explain
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
